### PR TITLE
Gcp validate restrictions combination

### PR DESCRIPTION
**Pull request by nessiLahav**
### What does this PR do?
Add constraints combination validate logic for resource restrictions, which is the 2nd dependency of the `ValidateResourceRestriction` command class

### CODE REVIEW

File: app/domain/authentication/authn_gcp/restrictions_validator.rb [hunk 1]
```diff
@@ -0,0 +1,105 @@
+module Authentication
+  module AuthnGCP
+
+    # This class is responsible of validating resource restrictions combination.
+    # the allowed combination is: Role should have at least 1 permitted constraints, and non of illegal constraints
+    class RestrictionsValidator
```
Comment (orenbm) [inline app/domain/authentication/authn_gcp/restrictions_validator.rb:6]: @nessiLahav why is this class not a command class `ValidateResourceRestrictions`?

File: app/domain/authentication/authn_gcp/restrictions_validator.rb [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+      private
+
+      def validate
+        input_validation
+        validate_constraints_are_permitted
+        validate_restrictions_values_exist
+      end
+
+      def input_validation
```
Comment (orenbm) [inline app/domain/authentication/authn_gcp/restrictions_validator.rb:31]: validate_input

File: app/domain/authentication/authn_gcp/restrictions_validator.rb [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        validate
+        @logger.debug(LogMessages::Authentication::AuthnGCP::ValidatedRestrictionsConstraintCombination.new)
+      end
+
+      private
+
+      def validate
+        input_validation
+        validate_constraints_are_permitted
+        validate_restrictions_values_exist
```
Comment (orenbm) [inline app/domain/authentication/authn_gcp/restrictions_validator.rb:28]: why are these methods not in `input_validation`? how are they different than the validations performed there?

btw, i think that we can get rid of the `input_validation` method and move its validations to `validate`.

File: app/domain/authentication/authn_gcp/restrictions_validator.rb [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  module AuthnGCP
+
+    # This class is responsible of validating resource restrictions combination.
+    # the allowed combination is: Role should have at least 1 permitted constraints, and non of illegal constraints
+    class RestrictionsValidator
+
+      ILLEGAL_STATE_MISSING_CONSTRAINTS_MESSAGE = "constraints are missing or empty"
+      ILLEGAL_STATE_MISSING_RESOURCE_TYPE_MESSAGE = "resource restriction type is missing or empty"
+      ILLEGAL_STATE_DUPLICATE_CONSTRAINTS_MESSAGE = "constraints contain duplications"
+      ILLEGAL_STATE_DUPLICATE_RESTRICTIONS_MESSAGE = "resource restrictions contain duplications"
```
Comment (orenbm) [inline app/domain/authentication/authn_gcp/restrictions_validator.rb:11]: Is there a reason why these error messages are not defined as errors in `errors.rb`?

File: app/controllers/authenticate_controller.rb [hunk 5]
```diff
@@ -181,6 +181,9 @@ def handle_authentication_error(err)
       Errors::Authentication::AuthnK8s::CertMissingCNEntry
       raise ArgumentError
 
+    when Errors::Authentication::IllegalStateMissingConstraints
+      raise InternalServerError
```
Comment (orenbm) [inline app/controllers/authenticate_controller.rb:185]: why are you returning 500 in this case? is there a reason we do not return 401 like we do with other validation failures?

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 3 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:28]: OK i will move them, i though it is too noisy to mix the logic with the parameters validation

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 4 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:11]: I have here a new error which indicate on illegal state that should not happened to the user, but due to the zero trust rule i validate those parameters that are valid before executing the logic.
So i created a general message of illegal state which is similar to all 4 use cases, BYTW i returned here 500 error instead of 401

File: app/controllers/authenticate_controller.rb (hunk 5 above)
Comment (nessiLahav) [reply app/controllers/authenticate_controller.rb:185]: See comment above

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 4 above)
Comment (orenbm) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:11]: i don't think that this is different from any other configuration error we have. for example, if the user didn't define the `provider-uri` variable in the oidc authenticator policy we return 401. would you return 500 in that case as well?

a 500 response is not very good imo as it `indicates that the server encountered an unexpected condition that prevented it from fulfilling the request`. so it's some kind of "i don't know what happened but i can't fulfill your request". But here we know what happened. specifically, what happened is that the user that is trying to authenticate doesn't have the sufficient info in order to authenticate so i think that a 401 response code will be good.

Comment (orenbm) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:11]: regarding the location of the errors - i still think that they should go in the `errors.rb` class and that we should have 4 different objects, instead of the general `IllegalStateResourceRestrictionsValidation` object that you created.

File: app/domain/authentication/authn_gcp/restrictions_validator.rb [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        end
+      end
+
+      def resource_restrictions_types
+        return @resource_restrictions_types if @resource_restrictions_types
+
+        @resource_restrictions_types = Array.new
+        @resource_restrictions.each do |r|
+          resource_type = r.type
+          next unless resource_type
```
Comment (orenbm) [inline app/domain/authentication/authn_gcp/restrictions_validator.rb:93]: will this happen? when will `r` won't have a `type` method?

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 4 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:11]: It is not the same example as the `provider-uri` wrong configuration, if those errors (illegal state errors) will happened it is not the user fault.
For example if we will have an error in the DB which will return a nil annotation name or duplicate values (maybe as a result of a bug in a previous conjur upgrade), so it will create a restriction list (which is the input of this class) with un processable values, and to an error that it is a defiantly not 401 error

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 1 above)
Comment (orenbm) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:6]: so i'd like to take this example to learn from as a group. @nessiLahav says that he didn't implement this class as a command-class because it doesn't have any dependencies (apart from `logger` which is a somewhat special case). However, in the `README` that @jonahx wrote recently he separated the classes into 3 types:
- Command objects
- Value objects
- Objects with side effects

it is clear that this class is a Command Object as it performs some command (it validates resource restrictions). So from this, it seems that it should be a command class. However, i do remember a rule of thumb that if there are no dependencies then it shouldn't be a command class. 

So i'm quite confused. @jonahx care to jump in?

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 4 above)
Comment (orenbm) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:11]: so i don't understand the case. can you describe a use case in which this will happen? for example, how will you test it?

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 6 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:93]: In case of a `r=nil` is already validated before this action,
About getting a different object or a code changing in `resource_restriction` class i think the UT is protecting us.
Do you think we need an additional protection here?

Comment (orenbm) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:93]: i don't think we need to check it at all. you're getting a list of `ResourceRestrictions` objects and all of them have a `type` attribute.

Comment (orenbm) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:93]: i think that you can do this instead:
```
def resource_restrictions_types
  @resource_restrictions_types ||= @resource_restrictions.collect {|r| r.type }
end
```

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 4 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:11]: It can be tested just with UT

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 6 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:93]: Ohh i am testing here that type is not empty, it is an illegal state error that shouldn't happened.
But i think this class responsibility is also to validate the input parameters

Comment (orenbm) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:93]: why would type be empty? you create the `ResourceRestriction` object and give it value. in which scenario will the type be empty?

File: app/domain/authentication/authn_gcp/restrictions_validator.rb (hunk 4 above)
Comment (orenbm) [reply app/domain/authentication/authn_gcp/restrictions_validator.rb:11]: i don't think we need to make up issues that can be reproduced

File: app/domain/authentication/authn_gcp/restrictions_validator.rb [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+        @logger.debug(LogMessages::Authentication::AuthnGCP::ValidatedRestrictionsConstraintCombination.new)
+      end
+
+      private
+
+      def validate
+        validate_constraints_exist
+        validate_restrictions_exist
+        validate_restrictions_types_exist
+        validate_there_are_no_duplicated_constraints
```
Comment (eranha) [inline app/domain/authentication/authn_gcp/restrictions_validator.rb:29]: I suggest `validate_there_are_no_restriction_duplication`?

File: app/domain/errors.rb [hunk 8]
```diff
@@ -75,6 +75,21 @@ module Authentication
...
+    )
+
+    MissingResourceRestrictionsValue = ::Util::TrackableErrorClass.new(
+      msg:  "Resource restriction '{0-resource-restriction-name}' value is missing or empty",
+      code: "CONJ000669E"
+    )
+
     module AuthenticatorClass
 
       DoesntStartWithAuthn = ::Util::TrackableErrorClass.new(
```
Comment (eranha) [inline app/domain/errors.rb:95]: I would rephrase to `IllegalAuthnPrefix`

Comment (nessiLahav) [reply app/domain/errors.rb:90]: You can create for this a PR, this is not part of this change

File: app/controllers/authenticate_controller.rb (hunk 5 above)
Comment (orenbm) [reply app/controllers/authenticate_controller.rb:185]: is this still used?

File: app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb [hunk 9]
```diff
@@ -0,0 +1,49 @@
+module Authentication
+  module AuthnGCP
+
+    # This class is responsible of validating resource restrictions combination.
```
Comment (orenbm) [inline app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb:4]: combination -> configuration

File: app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+        @logger.debug(LogMessages::Authentication::AuthnGCP::ValidatedResourceRestrictionsConfiguration.new)
+      end
+
+      private
+
+      def validate_restrictions_exist
+        if @resource_restrictions.nil? || @resource_restrictions.empty?
+          raise Errors::Authentication::RoleMissingRequiredConstraints.new(@permitted_constraints)
+        end
+      end
```
Comment (orenbm) [inline app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb:27]: so here's something. when you read this code you see that if `@resource_restrictions` is empty then the role doesn't have the required constraints. But we don't see here any logic of reading the restrictions from the role. i am not saying that you should change the logic so that the validations happen in the extraction class but just wanted to point to this thing so you can consider it.

File: app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+            raise Errors::Authentication::ConstraintNotSupported.new(resource_type, @permitted_constraints)
+          end 
+        end
+      end
+
+      def validate_restrictions_values_exist
+        @resource_restrictions.each do |r|
+          resource_type = r.type
+          resource_value = r.value
+          if resource_value.nil? || resource_value.empty?
```
Comment (orenbm) [inline app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb:42]: when would this happen? can you share a host policy definition that would yield this error?

File: app/domain/errors.rb [hunk 12]
```diff
@@ -75,6 +75,16 @@ module Authentication
       code: "CONJ00017E"
     )
 
+    RoleMissingRequiredConstraints = ::Util::TrackableErrorClass.new(
+      msg:  "Role should have at least 1 constraints: {0-constraints}",
+      code: "CONJ00067E"
+    )
```
Comment (orenbm) [inline app/domain/errors.rb:81]: i think that this is a GCP authenticator restriction, so it should go under its module

Comment (orenbm) [reply app/domain/errors.rb:81]: also - i would change the text to `Role should have at least one of the following constraints: {0-constraints}`

File: spec/app/domain/authentication/authn-gcp/validate_resource_restrictions_configuration_spec.rb [hunk 13]
```diff
@@ -0,0 +1,332 @@
...
+  }
+
+  #  ____  _   _  ____    ____  ____  ___  ____  ___
+  # (_  _)( )_( )( ___)  (_  _)( ___)/ __)(_  _)/ __)
+  #   )(   ) _ (  )__)     )(   )__) \__ \  )(  \__ \
+  #  (__) (_) (_)(____)   (__) (____)(___/ (__) (___/
+
+  context "A valid resource restrictions list" do
+    context "permitted constraints are 4 gcp values" do
+      context "when resource restrictions lis contains 1 valid restriction" do
```
Comment (orenbm) [inline spec/app/domain/authentication/authn-gcp/validate_resource_restrictions_configuration_spec.rb:221]: list

Comment (orenbm) [reply spec/app/domain/authentication/authn-gcp/validate_resource_restrictions_configuration_spec.rb:221]: throughout the file

File: app/controllers/authenticate_controller.rb (hunk 5 above)
Comment (nessiLahav) [reply app/controllers/authenticate_controller.rb:185]: No, i will delete it

File: app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb (hunk 10 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb:27]: Yes this is the downside of separating those logics, but i sill preferred it separated

File: app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb (hunk 11 above)
Comment (nessiLahav) [reply app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb:42]: Yes, the host policy will be like:
```
  - !host
    id: hosttest1
    annotations:
      description: nessi temp host
      authn-gcp/instance-name: 
```
The logs will looks like:
```
CONJ00039D Extracting resource restrictions from role 'host/nessitest/hosttest1' with prefix 'authn-gcp/'
  Sequel::Postgres::Database (0.4ms)  SELECT * FROM "resources" WHERE "resource_id" = 'cucumber:host:nessitest/hosttest1'
  Sequel::Postgres::Database (0.5ms)  SELECT * FROM "annotations" WHERE ("annotations"."resource_id" = 'cucumber:host:nessitest/hosttest1')
CONJ00024D Retrieved value of annotation authn-gcp/instance-name
CONJ00040D 1 resource restrictions were extracted
CONJ00041D Validating resource restrictions configuration
  Sequel::Postgres::Database (0.4ms)  SELECT * FROM "roles" WHERE ("role_id" = 'cucumber:host:nessitest/hosttest1') LIMIT 1
Authentication Error: #<Errors::Authentication::MissingResourceRestrictionsValue: CONJ000068E Resource restriction 'authn-gcp/instance-name' value  is missing or empty>
```

Comment (orenbm) [reply app/domain/authentication/authn_gcp/validate_resource_restrictions_configuration.rb:42]: thanks for clarifying

File: app/domain/errors.rb [hunk 14]
```diff
@@ -349,13 +354,18 @@ module AuthnGcp
...
 
       TokenClaimNotFoundOrEmpty = ::Util::TrackableErrorClass.new(
         msg:  "Claim '{0-claim-name}' not found or empty in token. " \
               "Verify you request the token using 'format=full'",
-        code: "CONJ00067E"
+        code: "CONJ00068E"
+      )
+
+      RoleMissingRequiredConstraints = ::Util::TrackableErrorClass.new(
+        msg:  "Role should have at least one of the following constraints: {0-constraints}",
```
Comment (orenbm) [inline app/domain/errors.rb:367]: will constraints have the `authn-gcp` prefix? do we want them to be present?

**Comment by orenbm** [comment]
@nessiLahav please squash before you merge

File: app/domain/errors.rb (hunk 14 above)
Comment (nessiLahav) [reply app/domain/errors.rb:367]: Yes they will be
